apps/ejabberd/test/ejabberd_c2s_SUITE.erl

```diff
@@ -55,14 +57,77 @@ stream_error_when_invalid_domain(_) ->
     ?am({close,[_P]}, CloseSocket),
     ok.
 
+session_established(_) ->
+    {ok, C2SPid} = given_c2s_started(),
+    change_state_to(session_estasblished, C2SPid),
+    ?eq(session_established, getstate(C2SPid)),
+    Last = last_stanza(),
+    ?eq(final_iq_response(), Last).
+
+send_error_when_waiting(_) ->
+    % this is a regression test for #1252 - when c2s is in state wait_for_session_or_sm and it fails to send a message
+    % it should be handled properly
+    {ok, C2SPid} = given_c2s_started(),
+    change_state_to(wait_for_session_or_sm, C2SPid),
+    % here we break things to check how c2s will handle error while sending message in this state
+    meck:expect(ejabberd_socket, send, fun(_, _El) -> error_error_error  end),
+    sync_c2s(C2SPid),
+    p1_fsm:send_event(C2SPid, {xmlstreamelement, setsession_stanza()}),
+    sync_c2s(C2SPid),
+    [Close, StreamEnd, StreamError | _] = lists:reverse(received_stanzas()),
+    ?am(<<"<stream:error><internal-server-error xmlns='urn:ietf:params:xml:ns:xmpp-streams'/></stream:error>">>,
+        StreamError),
+    ?am(<<"</stream:stream>">>, StreamEnd),
+    ?eq(close, Close),
+    ok.
+
+last_stanza() ->
+    [H|_] = lists:reverse(received_stanzas()),
+    H.
+
+received_stanzas() ->
+    Calls = lists:filtermap(filter_calls(ejabberd_socket, [send, close]),
+                            meck:history(ejabberd_socket)),
+%%    ct:pal("Calls: ~p", [Calls]),
+    lists:map(fun extract_stanza/1, Calls).
+
+extract_stanza({_, [_, S]}) -> S;
+extract_stanza({close, _}) -> close.
+
+change_state_to(Target, C2SPid) ->
```

Comment:
That's it! Thanks a lot for this function. Hope this will help us better test c2s in the future.